scripts/generate_matrix.py

```diff
@@ -0,0 +1,227 @@
+    )
+    np_minimum_dep = next(ver for ver in np_minimum_dep.split(",") if ">" in ver)
+    return Version.from_str(np_minimum_dep.replace(dependency, "").replace(">=", "").replace(">", ""))
+
+
+def get_python_versions(
+    min_version: Version | None = None,
+    max_version: Version | None = None,
+) -> list[Version]:
+    """
+    Fetch the list of available Python versions from the GitHub Actions Python Versions Manifest.
+
+    Args:
+        min_version (Version): The minimum version to include.
+        max_version (Version): The maximum version to include.
+        include_prerelease (bool): Whether to include pre-release versions.
+
+    Returns:
+        list[Version]: A list of available Python versions.
+
+    Raises:
+        urllib.error.URLError: If fetching data fails.
+
+    Examples:
+        >>> versions = get_python_versions()
+        >>> versions[0] >= Version(3, 6, 0)
+        True
+    """
+    data = fetch_json("https://raw.githubusercontent.com/actions/python-versions/main/versions-manifest.json")
+
+    versions = {}
+
+    for release in data[::-1]:
+        try:
+            version = Version.from_str(release["version"])
+        except ValueError:
+            # Skip pre-release versions
+            continue
+
+        if min_version and version < min_version:
+            continue
+        if max_version and version > max_version:
+            continue
+
+        versions[version.minor_version] = version
+
+    return sorted(versions.values())
+
+
+@lru_cache(maxsize=CACHE_SIZE)
+def fetch_json(url: str) -> dict | list:
+    """
+    Fetch JSON data from a URL with caching.
+
+    Args:
+        url (str): The URL to fetch.
+
+    Returns:
+        Union[Dict, List]: The parsed JSON data.
+
+    Raises:
+        urllib.error.URLError: If fetching data fails.
+
+    Examples:
+        >>> data = fetch_json("https://pypi.org/pypi/numpy/json")
+        >>> isinstance(data, dict)
+        True
+    """
+    try:
+        with urllib.request.urlopen(url) as response:  # noqa: S310
+            return json.loads(response.read().decode())
+    except urllib.error.URLError:
+        sys.exit(1)
+
+
+def get_package_versions(package_name: str, min_version: Version) -> list[str]:
+    """
+    Fetch available package versions from PyPI starting from the minimum version.
+
+    Args:
+        package_name (str): The name of the package on PyPI.
+        min_version_str (str): The minimum version string.
+
+    Returns:
+        List[str]: A sorted list of package versions as strings.
+
+    Examples:
+        >>> versions = get_package_versions("numpy", "1.21")
+        >>> "1.21.0" in versions
+        True
+    """
+    data = fetch_json(f"https://pypi.org/pypi/{package_name}/json")
+    releases = data.get("releases", {})
+    versions = {}
+    for version_str in releases:
+        if any(c in version_str for c in ["a", "b", "rc", "dev"]):
+            continue  # Skip pre-releases
```

Comment:
I'm not sure that's always what we want, given the upcoming `scipy=1.15.0rc1` that we're gonna be using